python/activator/middleware_interface.py

```diff
@@ -418,4 +498,40 @@ def run_pipeline(self, visit: Visit, exposure_ids: set) -> None:
         # If this is a fresh (local) repo, then types like calexp,
         # *Diff_diaSrcTable, etc. have not been registered.
         result = executor.run(register_dataset_types=True)
-        _log.info(f"Pipeline successfully run on {len(result)} quanta.")
+        _log.info(f"Pipeline successfully run on {len(result)} quanta for "
+                  f"detector {visit.detector} of {exposure_ids}.")
+
+
+def _query_missing_datasets(src_repo: Butler, dest_repo: Butler,
+                            *args, **kwargs) -> collections.abc.Iterable[lsst.daf.butler.DatasetRef]:
+    """Return datasets that are present in one repository but not another.
+
+    Parameters
+    ----------
+    src_repo : `lsst.daf.butler.Butler`
+        The repository in which a dataset must be present.
+    dest_repo : `lsst.daf.butler.Butler`
+        The repository in which a dataset must not be present.
+    *args, **kwargs
+        Parameters for describing the dataset query. They have the same
+        meanings as the parameters of `lsst.daf.butler.Registry.queryDatasets`.
+
+    Returns
+    -------
+    datasets : iterable [`lsst.daf.butler.DatasetRef`]
+        The datasets that exist in ``src_repo`` but not ``dest_repo``.
+    """
+    try:
+        # TODO: storing this set in memory may be a performance bottleneck.
+        # In general, expect dest_repo to have more datasets than src_repo.
```

Comment:
Does anything go wrong if that expectation is violated?

Reply:
Well, it's violated every time we use a fresh repo!

I meant this sentence to be a continuation of the previous one: not only are we materializing one of the sets, it's the bigger one (given an old enough repo).

Reply:
I don't know of any algorithm for a set difference that's streaming on both sets, unless they're sorted (which `queryDatasets` does *not* guarantee), and it sounds like [nobody else does, either](https://stackoverflow.com/questions/26547286/how-to-get-the-symmetric-difference-between-two-streams-in-java-8).

Reply:
Ah, ok. Maybe tweak the comment a bit with the above information somehow? It wasn't entirely clear to me that the second line was a continuation of that though.

Reply:
Actually, I deleted it. I think it is impossible for the `dest_repo` results to be bigger than the `src_repo` ones (sine the two queries use the same arguments, that would imply that there is something in `dest_repo` calibs/templates/refcats that did not come from `src_repo`).